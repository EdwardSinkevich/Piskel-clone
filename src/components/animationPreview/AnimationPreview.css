.animated-preview-container {
  padding-left: 10px;
  position: relative;
  width: 200px;
}

.canvas-container-wrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #333;
}

.animated-preview-canvas-container {
  position: relative;
  height: 200px;
  width: 200px;
  margin: 0px;
}

:fullscreen .animated-preview-canvas-container {
  width: 90vh;
  height: 90vh;
}

:fullscreen .full-screen-btn {
  display: none;
}

.animation-player {
  height: 100%;
  width: 100%;
  position: relative;
  background-position: center;
  background-size: cover;
}

.display-fps-container {
  height: 30px;
  display: flex;
  align-items: center;
  align-self: center;
}

.display-fps {
  min-width: 45px;
  margin: 0 5px;
  font-size: 12px;
  color: #aaa;
}

.range-fps {
  width: 100px;
}

.full-screen-btn {
  position: relative;
  padding: 5px 5px;
}

.full-screen-tooltip {
  left: -19px;
  top: 26px;
}
