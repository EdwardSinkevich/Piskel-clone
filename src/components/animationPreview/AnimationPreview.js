import './AnimationPreview.css';

export default class AnimationPreview {
  constructor(frames) {
    this.animatedPreview = null;
    this.animationPlayer = null;
    this.frames = frames;
    this.imgsArray = this.frames.imgsArray;
    this.intervalId = null;
    this.intervalCounter = 0;
    this.fps = 1000;
    this.rangeFps = null;
    this.fullScreenBtn = null;
  }

  startAnimation() {
    this.intervalId = setInterval(() => {
      if (this.imgsArray.length >= 1 && this.intervalCounter < this.imgsArray.length) {
        this.animationPlayer.style.backgroundImage = `url(${this.imgsArray[this.intervalCounter].src})`;
        this.intervalCounter += 1;
      }

      if (this.intervalCounter >= this.imgsArray.length) {
        if (this.imgsArray[this.intervalCounter - 1]) {
          this.animationPlayer.style.backgroundImage = `url(${this.imgsArray[this.intervalCounter - 1].src})`;
        }
        this.intervalCounter = 0;
      }
    }, this.fps);
  }

  redrawingOnAnimationPlayer() {
    if (this.intervalCounter === 0) {
      document.getElementById('animation-player').style.backgroundImage = `url(${this.frames.mainCanvasObj.currentMainCanvasImg.src})`;
    }
  }

  changeFps() {
    this.fps = 1000 / this.rangeFps.value;
    const displayFps = document.getElementById('display-fps');

    displayFps.innerText = `${this.rangeFps.value} FPS`;

    if (this.fps === Infinity) {
      displayFps.innerText = `${this.rangeFps.value} FPS`;

      this.intervalCounter = 0;

      clearInterval(this.intervalId);

      this.redrawingOnAnimationPlayer();
      return;
    }
    clearInterval(this.intervalId);
    this.startAnimation();
  }

  launchToFullscreen() {
    const animationPlayerContainer = document.getElementById('canvas-container-wrapper');
    animationPlayerContainer.webkitRequestFullscreen();
  }

  addListeners() {
    const redrawingOnAnimationPlayerWrap = this.redrawingOnAnimationPlayer.bind(this);
    this.frames.mainCanvas.addEventListener('mouseup', redrawingOnAnimationPlayerWrap);
    this.frames.mainCanvas.addEventListener('mouseleave', redrawingOnAnimationPlayerWrap);

    const changeFpsWrap = this.changeFps.bind(this);
    this.rangeFps.addEventListener('input', changeFpsWrap);

    const launchToFullscreenWrap = this.launchToFullscreen.bind(this);
    this.fullScreenBtn.addEventListener('click', launchToFullscreenWrap);
  }

  render() {
    this.animatedPreview = document.getElementById('animated-preview-container');
    this.animatedPreview.innerHTML = `
    <div id="canvas-container-wrapper" class="canvas-container-wrapper">
      <div id="animated-preview-canvas-container" class="animated-preview-canvas-container">
        <div class="canvas-background"></div>
        <div id="animation-player" class="animation-player"></div>
      </div>
      <div class="display-fps-container">
        <button id="full-screen-btn" class="nav-btn full-screen-btn tooltip-action">Full
        <span class="tooltip full-screen-tooltip">Press F</span>
        </button>
        <span id="display-fps" class="display-fps">1 FPS</span>
        <input id="range-fps" class="range-fps" type="range" min="0" step="1" value="1" max="24">
      </div>
    </div>`;

    this.animationPlayer = document.getElementById('animation-player');
    this.rangeFps = document.getElementById('range-fps');
    this.fullScreenBtn = document.getElementById('full-screen-btn');

    this.addListeners();

    this.startAnimation();
  }
}
