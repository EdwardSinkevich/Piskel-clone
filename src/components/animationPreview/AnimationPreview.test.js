import AnimationPreview from './AnimationPreview';

describe('AnimationPreview.constructor', () => {
  it('Should assign this.frames to frames', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    expect(animationPreview.frames).toEqual(testObj);
  });
});

describe('AnimationPreview.startAnimation', () => {
  it('Should assign this.intervalId', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.startAnimation();
    expect(animationPreview.intervalId).not.toBeNull();
    clearInterval(animationPreview.intervalId);
  });
});

describe('AnimationPreview.redrawingOnAnimationPlayer', () => {
  it('Should assign animation player backgroundImage to this.frames.mainCanvasObj.currentMainCanvasImg.src', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.intervalCounter = 0;

    const animationPlayer = document.createElement('div');
    animationPlayer.id = 'animation-player';
    document.body.appendChild(animationPlayer);

    animationPreview.frames = {
      mainCanvasObj: {
        currentMainCanvasImg: {
          src: 'testUrl',
        },
      },
    };
    animationPreview.redrawingOnAnimationPlayer();
    const animationPlayerAfter = document.getElementById('animation-player');
    const animationPlayerBackgroundImage = getComputedStyle(animationPlayerAfter).getPropertyValue('background-image');
    expect(animationPlayerBackgroundImage).toEqual(`url(${animationPreview.frames.mainCanvasObj.currentMainCanvasImg.src})`);
  });

  it('Should not assign animation player backgroundImage to this.frames.mainCanvasObj.currentMainCanvasImg.src if this.intervalCounter not equals 0', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.intervalCounter = 1;

    const animationPlayer = document.createElement('div');
    animationPlayer.id = 'animation-player';
    document.body.appendChild(animationPlayer);

    animationPreview.frames = {
      mainCanvasObj: {
        currentMainCanvasImg: {
          src: 'testUrlIfNotZero',
        },
      },
    };
    animationPreview.redrawingOnAnimationPlayer();
    const animationPlayerAfter = document.getElementById('animation-player');
    const animationPlayerBackgroundImage = getComputedStyle(animationPlayerAfter).getPropertyValue('background-image');
    expect(animationPlayerBackgroundImage).not.toEqual(`url(${animationPreview.frames.mainCanvasObj.currentMainCanvasImg.src})`);
  });
});

describe('AnimationPreview.changeFps', () => {
  const displayFps = document.createElement('div');
  displayFps.id = 'display-fps';
  document.body.appendChild(displayFps);

  it('Should set fps value in element with id equals display-fps', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.rangeFps = {
      value: 1,
    };

    animationPreview.changeFps();
    expect(displayFps.innerText).toEqual(`${animationPreview.rangeFps.value} FPS`);
  });

  it('Should set fps value in element with id equals display-fps to 0 if this.rangeFps.value === 0', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.rangeFps = {
      value: 0,
    };
    animationPreview.redrawingOnAnimationPlayer = () => {};

    animationPreview.changeFps();
    expect(displayFps.innerText).toEqual(`${animationPreview.rangeFps.value} FPS`);
  });
});

describe('AnimationPreview.launchToFullscreen', () => {
  it('Should call webkitRequestFullscreen', () => {
    const myMock = jest.fn();
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);

    const animationPlayerContainer = document.createElement('div');
    animationPlayerContainer.id = 'canvas-container-wrapper';
    document.body.appendChild(animationPlayerContainer);

    animationPlayerContainer.webkitRequestFullscreen = myMock;
    animationPreview.launchToFullscreen();
    expect(myMock.mock.invocationCallOrder.length).toBe(1);
  });
});

describe('AnimationPreview.render', () => {
  it('Should render correctly', () => {
    const testObj = {};
    const animationPreview = new AnimationPreview(testObj);
    animationPreview.addListeners = () => {};

    const animatedPreview = document.createElement('div');
    animatedPreview.id = 'animated-preview-container';
    document.body.appendChild(animatedPreview);
    animationPreview.render();
    expect(document.getElementById('animated-preview-container').innerHTML).toMatchSnapshot();
  });
});
