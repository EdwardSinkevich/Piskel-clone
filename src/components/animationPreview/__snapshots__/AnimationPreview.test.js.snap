// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimationPreview.render Should render correctly 1`] = `
"
    <div id=\\"canvas-container-wrapper\\" class=\\"canvas-container-wrapper\\">
      <div id=\\"animated-preview-canvas-container\\" class=\\"animated-preview-canvas-container\\">
        <div class=\\"canvas-background\\"></div>
        <div id=\\"animation-player\\" class=\\"animation-player\\"></div>
      </div>
      <div class=\\"display-fps-container\\">
        <button id=\\"full-screen-btn\\" class=\\"nav-btn full-screen-btn tooltip-action\\">Full
        <span class=\\"tooltip full-screen-tooltip\\">Press F</span>
        </button>
        <span id=\\"display-fps\\" class=\\"display-fps\\">1 FPS</span>
        <input id=\\"range-fps\\" class=\\"range-fps\\" type=\\"range\\" min=\\"0\\" step=\\"1\\" value=\\"1\\" max=\\"24\\">
      </div>
    </div>"
`;
